pkg/controller/certificatesigningrequest/ipsec_csr_signing_controller.go

```diff
@@ -0,0 +1,486 @@
+	}
+	if len(certs) != 1 {
+		return nil, fmt.Errorf("expect a single certificate, got %d", len(certs))
+	}
+	return certs[0], nil
+}
+
+// NewIPsecCSRSigningController returns a new *IPsecCSRSigningController.
+func NewIPsecCSRSigningController(client clientset.Interface,
+	csrInformer csrinformer.CertificateSigningRequestInformer,
+	selfSignedCA bool,
+) (*IPsecCSRSigningController, error) {
+
+	caConfigMapInformer := corev1informers.NewFilteredConfigMapInformer(client, env.GetAntreaNamespace(), 12*time.Hour, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}, func(listOptions *v1.ListOptions) {
+		listOptions.FieldSelector = fields.OneTermEqualSelector("metadata.name", ipsecRootCAName).String()
+	})
+
+	configmapLister := corev1listers.NewConfigMapLister(caConfigMapInformer.GetIndexer())
+
+	c := &IPsecCSRSigningController{
+		client:                client,
+		csrLister:             csrInformer.Lister(),
+		csrListerSynced:       csrInformer.Informer().HasSynced,
+		configMapInformer:     caConfigMapInformer,
+		configMapLister:       configmapLister,
+		configMapListerSynced: caConfigMapInformer.HasSynced,
+		selfSignedCA:          selfSignedCA,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "certificateSigningRequest"),
+		fixturesQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "certificateSigningRequest"),
+	}
+
+	csrInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: c.enqueueCertificateSigningRequest,
+			UpdateFunc: func(old, cur interface{}) {
+				c.enqueueCertificateSigningRequest(cur)
+			},
+			DeleteFunc: c.enqueueCertificateSigningRequest,
+		},
+		resyncPeriod,
+	)
+
+	caConfigMapInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.fixturesQueue.Add(workerItemKey)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				c.fixturesQueue.Add(workerItemKey)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.fixturesQueue.Add(workerItemKey)
+			},
+		},
+		resyncPeriod,
+	)
+
+	if err := c.syncRootCertificateAndKey(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run begins watching and syncing of the IPsecCSRSigningController.
+func (c *IPsecCSRSigningController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", ipsecCSRSigningControllerName)
+	defer klog.Infof("Shutting down %s", ipsecCSRSigningControllerName)
+
+	go c.configMapInformer.Run(stopCh)
+
+	cacheSyncs := []cache.InformerSynced{c.csrListerSynced, c.configMapListerSynced}
+	if !cache.WaitForNamedCacheSync(ipsecCSRSigningControllerName, stopCh, cacheSyncs...) {
+		return
+	}
+
+	go wait.Until(c.fixturesWorker, time.Second, stopCh)
+
+	go wait.Until(func() {
+		if err := c.watchSecretChanges(stopCh); err != nil {
```

Comment:
I think it should only watch secret when selfSignedCA is false.

Reply:
But if selfSignedCA is set to True, does it need to take action when users change CA?

Reply:
Do you mean that agents need to rotate the certificates once users change CA? I think it is the same for rotating self-signed CA. We can watch for CA file changes in Agents and request new certificates if necessary. Tweaks for the `antrea-ipsec` container will also be needed to reload the new CA.

Reply:
No, I mean when selfSignedCA is set to True, which means antrea will generate a CA itself, why it needs to watch the secret which is supposed to be used when selfSignedCA is set to False?

Reply:
I think it would be helpful if user wants to rotate self-signed CA manually by just deleting the Secret. But we do not fully support this right now.

Reply:
Got it, the use case makes sense to me